Pull Rewrite asm in C into master branch

kevana opened the pull request.

kevana committed:
Initial include

Will not work.

kevana committed:
Finished adding DMX receiving code

Fixed obvious infinite loop in receiveDMX(), need to check for possible
infinite loop at line 150 in WaitForData.
Added setupDMX() for initial DMX input setup.
Cleaned up misc. errors preventing compilation. Still does not compile
in MPLabX, some pragmas throw errors and the _asm segments are not
being interpreted as such by the compiler.

kevana committed:
Updated Readme language (hopefully for clarity)

kevana committed:
synced README from master branch

kevana committed:
checked out README.md from master branch (for real this time)

Had not updated master branch, grabbed old README as a result. Should
be correct now.

kevana committed:
Revert "checked out README.md from master branch (for real this time)"

This reverts commit 97bff85c7b1f1e97ff4c4eee1efd191a818eecdb.

kevana committed:
Revert "synced README from master branch"

This reverts commit 7d753bee09191f9e37df97d42031e1796d2d43df.

kevana committed:
Merged master into Add-inline-ASM-to-receive-DMX

kevana committed:
Fixed formatting..... Again

kevana committed:
Bytes sent now flexible, FSR0 no longer used

fixes #5
fixes #8
Removed testing initialization values.

kevana committed:
Resolved conflicts from pulling master into this branch

kevana committed:
Added C code for receiveDMX()

All ASM code commented out and equivalent functionality in C
implemented. Program flow was not altered (e.g. label/goto still used)

taylortrimble commented:
Can we merge master into this so the diff I have to look at is littler? @aterlumen

kevana committed:
Added C for setupDMX function

kevana committed:
Replaced hard tabs with spaces and removed a break

Textwrangler default was not to auto-expand tabs.
break was remnant from a previous loop, removed

kevana committed:
Refactoring….

Explicitly defined DMXStartCode,
Removed goto GotBreak
Fixed logic at 120
Added various comments
Refactored 170-181 to remove a goto. Considering making a separate
function (Start a C DMX library?) for this to remove the other goto.
Added newline to end of file

kevana committed:
Added comments for clarity

kevana committed:
Created new file for DMX PIC functions

First draft using a switch statement to simulate a FSM. I don't have my
state diagram or a compiler with me so there's probably quite a few
errors, but it's a start.

@tylrtrmbl : Thoughts on setupDMX? I'm not sure how much of this should
be included in a library-like function vs. in the main file's setup
function.

taylortrimble commented on DMXlib.h:
```diff
@@ -0,0 +1,97 @@
+ */
+
+// Variables
+char RxBuffer[512];
```
Lets do:

``` C
const int DMXBufferSize = 512;
char DMXBuffer[DXMBufferSize];
```

taylortrimble commented on DMXlib.h:
```diff
@@ -0,0 +1,97 @@
+
+// Variables
+char RxBuffer[512];
+static int numReceivedBytes; //16-bit counter
```
`static` in a global variable doesn't make much sense! Lets use `int DMXBytesReceived`.

taylortrimble commented on DMXlib.h:
```diff
@@ -0,0 +1,97 @@
+// Variables
+char RxBuffer[512];
+static int numReceivedBytes; //16-bit counter
+char inputBuffer; //used to read RCREG to clear error conditions
```
`char DMXInputBuffer;`

taylortrimble commented on DMXlib.h:
```diff
@@ -0,0 +1,97 @@
+char inputBuffer; //used to read RCREG to clear error conditions
+const int DMXStartCode = 0x00;
+enum {
+    MainLoop, WaitBreak, GotBreak, WaitForStart, WaitForData, MoveData, Done
```
Can we do:

``` C
enum {
    DMXMainLoop, DMXWaitBreak, DMXGotBreak, DMXWaitForStart, DMXWaitForData, DMXMoveData, DMXDone
} DMXState;
```

:question: 
It's a little more verbose, but more framework-y styled.

taylortrimble commented on DMXlib.h:
```diff
@@ -0,0 +1,97 @@
+    MainLoop, WaitBreak, GotBreak, WaitForStart, WaitForData, MoveData, Done
+} state;
+
+void setupDMX(void)
```
`void DMXSetup(void)`

taylortrimble commented on DMXlib.h:
```diff
@@ -0,0 +1,97 @@
+void setupDMX(void)
+{
+    OSCTUNEbits.PLLEN = 1; //Set PLL on. Not sure why this isn't in the main setup function
+    for (int i = 0; i < 512; i++) { //Clear the receive buffer 
```
`512` :arrow_right: `DMXBufferSize`

taylortrimble commented:
Holy :poop: 5ae84f696249b356cfafcaa1d06f362bfcd6d76c is one of the most beautiful commits I've ever seen. That's some **_sexy_** code!

I've added a few initial line comments on style, along the lines of:

```
const int DMXConstantName;
int DMXGlobalName;
void DMXFunctionName(void) {}
```

To make it more framework-like. I'll go over it for accuracy later. But holy shit. Nice work @aterlumen!

taylortrimble commented:
Also, maybe we could move the code out of the header file? :stuck_out_tongue_winking_eye:

kevana committed:
Refactoring, stylistic changes

kevana committed:
Fixing logic errors

Removed unnecessary state, fixed misc. logic errors.
Fixed incorrect comments
Edited function name for consistency.
Debating whether to remove DMXGotBreak state

kevana committed:
Update to use DMXlib.h

Ripped out internal functions

kevana commented:
I think unless @tylrtrmbl finds any more logic errors in DMXlib.h this is ready to pull into master.

kevana committed:
Header file conventions

taylortrimble commented on DMXlib.c:
```diff
@@ -0,0 +1,101 @@
+{
+    OSCTUNEbits.PLLEN = 1; //Set PLL on. Not sure why this isn't in the main setup function
+    for (int i = 0; i < DMXBufferSize; i++) { //Clear the receive buffer 
+    	DMXBuffer[i] = 0;
```
Alignment; also on line 26, 27, 30

taylortrimble commented on DMXlib.c:
```diff
@@ -0,0 +1,101 @@
+char DMXBuffer[DMXBufferSize];
+int DMXBytesReceived; //16-bit counter
+char DMXInputBuffer; //used to read RCREG to clear error conditions
+const int DMXStartCode = 0x00;
```
Can we have separate

```
// Constants
...

// Variables
...
```

:question:

taylortrimble commented on DMXlib.c:
```diff
@@ -0,0 +1,101 @@
+
+void DMXSetup(void)
+{
+    OSCTUNEbits.PLLEN = 1; //Set PLL on. Not sure why this isn't in the main setup function
```
Let's move this there.

taylortrimble commented on DMXlib.h:
```diff
@@ -0,0 +1,16 @@
+ */
+ 
+ // Variables
+const int DMXBufferSize = 512;
```
We actually don't want this or line 11 declared twice; we should remove them from the `.c` file.

taylortrimble commented on DMXlib.c:
```diff
@@ -0,0 +1,101 @@
+ *
+ * Author: Kevan Ahlquist (@aterlumen)
+ */
+
```
`DMXlib.c` should `#include "DMXlib.h"`.

taylortrimble commented:
This is one of the best code ports I've ever seen. Hurrah! after all the stylistic, spacing, and other fixes I mentioned in the last diff and 51d5916bf1357f2d695b44270cfc423e765bc8fd are done, it's good to pull!
:shipit:

kevana committed:
Removed old LED code, Stylistic changes

Removed a logic state in DMXReceive, changed some styling.

kevana committed:
Reorganizing variables between .h and .c files, Indentation

Reorganized variables, moved around variables between .c and .h files,
updated formatting.

kevana commented:
@tylrtrmbl One last look? I think we're good to go here.

taylortrimble commented:
:sparkles:
